// aleo-hooks/dist/index.es.js
var Pr = Object.defineProperty, Dr = Object.defineProperties;
var Nr = Object.getOwnPropertyDescriptors;
var Ge = Object.getOwnPropertySymbols;
var jr = Object.prototype.hasOwnProperty, kr = Object.prototype.propertyIsEnumerable;
var Re = (t, n, o) => n in t ? Pr(t, n, { enumerable: !0, configurable: !0, writable: !0, value: o }) : t[n] = o, He = (t, n) => {
  for (var o in n || (n = {}))
    jr.call(n, o) && Re(t, o, n[o]);
  if (Ge)
    for (var o of Ge(n))
      kr.call(n, o) && Re(t, o, n[o]);
  return t;
}, Xe = (t, n) => Dr(t, Nr(n));
var Y = (t, n, o) => (Re(t, typeof n != "symbol" ? n + "" : n, o), o);
var $ = (t, n, o) => new Promise((u, p) => {
  var d = (s) => {
    try {
      f(o.next(s));
    } catch (i) {
      p(i);
    }
  }, v = (s) => {
    try {
      f(o.throw(s));
    } catch (i) {
      p(i);
    }
  }, f = (s) => s.done ? u(s.value) : Promise.resolve(s.value).then(d, v);
  f((o = o.apply(t, n)).next());
});
import rr, { useState as R, useRef as Se, useEffect as K, useMemo as A, createContext as Ir, useContext as Fr, useCallback as I } from "react";
var Te = { exports: {} }, ne = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Qe;
function Ar() {
  if (Qe)
    return ne;
  Qe = 1;
  var t = rr, n = Symbol.for("react.element"), o = Symbol.for("react.fragment"), u = Object.prototype.hasOwnProperty, p = t.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, d = { key: !0, ref: !0, __self: !0, __source: !0 };
  function v(f, s, i) {
    var a, c = {}, w = null, y = null;
    i !== void 0 && (w = "" + i), s.key !== void 0 && (w = "" + s.key), s.ref !== void 0 && (y = s.ref);
    for (a in s)
      u.call(s, a) && !d.hasOwnProperty(a) && (c[a] = s[a]);
    if (f && f.defaultProps)
      for (a in s = f.defaultProps, s)
        c[a] === void 0 && (c[a] = s[a]);
    return { $$typeof: n, type: f, key: w, ref: y, props: c, _owner: p.current };
  }
  return ne.Fragment = o, ne.jsx = v, ne.jsxs = v, ne;
}
var ae = {};
/**
 * @license React
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Ze;
function Lr() {
  return Ze || (Ze = 1, process.env.NODE_ENV !== "production" && function() {
    var t = rr, n = Symbol.for("react.element"), o = Symbol.for("react.portal"), u = Symbol.for("react.fragment"), p = Symbol.for("react.strict_mode"), d = Symbol.for("react.profiler"), v = Symbol.for("react.provider"), f = Symbol.for("react.context"), s = Symbol.for("react.forward_ref"), i = Symbol.for("react.suspense"), a = Symbol.for("react.suspense_list"), c = Symbol.for("react.memo"), w = Symbol.for("react.lazy"), y = Symbol.for("react.offscreen"), m = Symbol.iterator, T = "@@iterator";
    function g(e) {
      if (e === null || typeof e != "object")
        return null;
      var r = m && e[m] || e[T];
      return typeof r == "function" ? r : null;
    }
    var N = t.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    function S(e) {
      {
        for (var r = arguments.length, l = new Array(r > 1 ? r - 1 : 0), h = 1; h < r; h++)
          l[h - 1] = arguments[h];
        C("error", e, l);
      }
    }
    function C(e, r, l) {
      {
        var h = N.ReactDebugCurrentFrame, b = h.getStackAddendum();
        b !== "" && (r += "%s", l = l.concat([b]));
        var x = l.map(function(_) {
          return String(_);
        });
        x.unshift("Warning: " + r), Function.prototype.apply.call(console[e], console, x);
      }
    }
    var q = !1, U = !1, se = !1, ee = !1, re = !1, J;
    J = Symbol.for("react.module.reference");
    function O(e) {
      return !!(typeof e == "string" || typeof e == "function" || e === u || e === d || re || e === p || e === i || e === a || ee || e === y || q || U || se || typeof e == "object" && e !== null && (e.$$typeof === w || e.$$typeof === c || e.$$typeof === v || e.$$typeof === f || e.$$typeof === s || // This needs to include all possible module reference object
      // types supported by any Flight configuration anywhere since
      // we don't know which Flight build this will end up being used
      // with.
      e.$$typeof === J || e.getModuleId !== void 0));
    }
    function L(e, r, l) {
      var h = e.displayName;
      if (h)
        return h;
      var b = r.displayName || r.name || "";
      return b !== "" ? l + "(" + b + ")" : l;
    }
    function W(e) {
      return e.displayName || "Context";
    }
    function j(e) {
      if (e == null)
        return null;
      if (typeof e.tag == "number" && S("Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."), typeof e == "function")
        return e.displayName || e.name || null;
      if (typeof e == "string")
        return e;
      switch (e) {
        case u:
          return "Fragment";
        case o:
          return "Portal";
        case d:
          return "Profiler";
        case p:
          return "StrictMode";
        case i:
          return "Suspense";
        case a:
          return "SuspenseList";
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case f:
            var r = e;
            return W(r) + ".Consumer";
          case v:
            var l = e;
            return W(l._context) + ".Provider";
          case s:
            return L(e, e.render, "ForwardRef");
          case c:
            var h = e.displayName || null;
            return h !== null ? h : j(e.type) || "Memo";
          case w: {
            var b = e, x = b._payload, _ = b._init;
            try {
              return j(_(x));
            } catch (E) {
              return null;
            }
          }
        }
      return null;
    }
    var F = Object.assign, H = 0, xe, Oe, Pe, De, Ne, je, ke;
    function Ie() {
    }
    Ie.__reactDisabledLog = !0;
    function ar() {
      {
        if (H === 0) {
          xe = console.log, Oe = console.info, Pe = console.warn, De = console.error, Ne = console.group, je = console.groupCollapsed, ke = console.groupEnd;
          var e = {
            configurable: !0,
            enumerable: !0,
            value: Ie,
            writable: !0
          };
          Object.defineProperties(console, {
            info: e,
            log: e,
            warn: e,
            error: e,
            group: e,
            groupCollapsed: e,
            groupEnd: e
          });
        }
        H++;
      }
    }
    function or() {
      {
        if (H--, H === 0) {
          var e = {
            configurable: !0,
            enumerable: !0,
            writable: !0
          };
          Object.defineProperties(console, {
            log: F({}, e, {
              value: xe
            }),
            info: F({}, e, {
              value: Oe
            }),
            warn: F({}, e, {
              value: Pe
            }),
            error: F({}, e, {
              value: De
            }),
            group: F({}, e, {
              value: Ne
            }),
            groupCollapsed: F({}, e, {
              value: je
            }),
            groupEnd: F({}, e, {
              value: ke
            })
          });
        }
        H < 0 && S("disabledDepth fell below zero. This is a bug in React. Please file an issue.");
      }
    }
    var ye = N.ReactCurrentDispatcher, he;
    function ie(e, r, l) {
      {
        if (he === void 0)
          try {
            throw Error();
          } catch (b) {
            var h = b.stack.trim().match(/\n( *(at )?)/);
            he = h && h[1] || "";
          }
        return `
` + he + e;
      }
    }
    var ge = !1, ce;
    {
      var sr = typeof WeakMap == "function" ? WeakMap : Map;
      ce = new sr();
    }
    function Fe(e, r) {
      if (!e || ge)
        return "";
      {
        var l = ce.get(e);
        if (l !== void 0)
          return l;
      }
      var h;
      ge = !0;
      var b = Error.prepareStackTrace;
      Error.prepareStackTrace = void 0;
      var x;
      x = ye.current, ye.current = null, ar();
      try {
        if (r) {
          var _ = function() {
            throw Error();
          };
          if (Object.defineProperty(_.prototype, "props", {
            set: function() {
              throw Error();
            }
          }), typeof Reflect == "object" && Reflect.construct) {
            try {
              Reflect.construct(_, []);
            } catch (B) {
              h = B;
            }
            Reflect.construct(e, [], _);
          } else {
            try {
              _.call();
            } catch (B) {
              h = B;
            }
            e.call(_.prototype);
          }
        } else {
          try {
            throw Error();
          } catch (B) {
            h = B;
          }
          e();
        }
      } catch (B) {
        if (B && h && typeof B.stack == "string") {
          for (var E = B.stack.split(`
`), k = h.stack.split(`
`), P = E.length - 1, D = k.length - 1; P >= 1 && D >= 0 && E[P] !== k[D]; )
            D--;
          for (; P >= 1 && D >= 0; P--, D--)
            if (E[P] !== k[D]) {
              if (P !== 1 || D !== 1)
                do
                  if (P--, D--, D < 0 || E[P] !== k[D]) {
                    var M = `
` + E[P].replace(" at new ", " at ");
                    return e.displayName && M.includes("<anonymous>") && (M = M.replace("<anonymous>", e.displayName)), typeof e == "function" && ce.set(e, M), M;
                  }
                while (P >= 1 && D >= 0);
              break;
            }
        }
      } finally {
        ge = !1, ye.current = x, or(), Error.prepareStackTrace = b;
      }
      var Z = e ? e.displayName || e.name : "", Je = Z ? ie(Z) : "";
      return typeof e == "function" && ce.set(e, Je), Je;
    }
    function ir(e, r, l) {
      return Fe(e, !1);
    }
    function cr(e) {
      var r = e.prototype;
      return !!(r && r.isReactComponent);
    }
    function le(e, r, l) {
      if (e == null)
        return "";
      if (typeof e == "function")
        return Fe(e, cr(e));
      if (typeof e == "string")
        return ie(e);
      switch (e) {
        case i:
          return ie("Suspense");
        case a:
          return ie("SuspenseList");
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case s:
            return ir(e.render);
          case c:
            return le(e.type, r, l);
          case w: {
            var h = e, b = h._payload, x = h._init;
            try {
              return le(x(b), r, l);
            } catch (_) {
            }
          }
        }
      return "";
    }
    var ue = Object.prototype.hasOwnProperty, Ae = {}, Le = N.ReactDebugCurrentFrame;
    function fe(e) {
      if (e) {
        var r = e._owner, l = le(e.type, e._source, r ? r.type : null);
        Le.setExtraStackFrame(l);
      } else
        Le.setExtraStackFrame(null);
    }
    function lr(e, r, l, h, b) {
      {
        var x = Function.call.bind(ue);
        for (var _ in e)
          if (x(e, _)) {
            var E = void 0;
            try {
              if (typeof e[_] != "function") {
                var k = Error((h || "React class") + ": " + l + " type `" + _ + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof e[_] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.");
                throw k.name = "Invariant Violation", k;
              }
              E = e[_](r, _, h, l, null, "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED");
            } catch (P) {
              E = P;
            }
            E && !(E instanceof Error) && (fe(b), S("%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).", h || "React class", l, _, typeof E), fe(null)), E instanceof Error && !(E.message in Ae) && (Ae[E.message] = !0, fe(b), S("Failed %s type: %s", l, E.message), fe(null));
          }
      }
    }
    var ur = Array.isArray;
    function we(e) {
      return ur(e);
    }
    function fr(e) {
      {
        var r = typeof Symbol == "function" && Symbol.toStringTag, l = r && e[Symbol.toStringTag] || e.constructor.name || "Object";
        return l;
      }
    }
    function dr(e) {
      try {
        return We(e), !1;
      } catch (r) {
        return !0;
      }
    }
    function We(e) {
      return "" + e;
    }
    function Ye(e) {
      if (dr(e))
        return S("The provided key is an unsupported type %s. This value must be coerced to a string before before using it here.", fr(e)), We(e);
    }
    var te = N.ReactCurrentOwner, vr = {
      key: !0,
      ref: !0,
      __self: !0,
      __source: !0
    }, $e, Ue, me;
    me = {};
    function pr(e) {
      if (ue.call(e, "ref")) {
        var r = Object.getOwnPropertyDescriptor(e, "ref").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.ref !== void 0;
    }
    function yr(e) {
      if (ue.call(e, "key")) {
        var r = Object.getOwnPropertyDescriptor(e, "key").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.key !== void 0;
    }
    function hr(e, r) {
      if (typeof e.ref == "string" && te.current && r && te.current.stateNode !== r) {
        var l = j(te.current.type);
        me[l] || (S('Component "%s" contains the string ref "%s". Support for string refs will be removed in a future major release. This case cannot be automatically converted to an arrow function. We ask you to manually fix this case by using useRef() or createRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref', j(te.current.type), e.ref), me[l] = !0);
      }
    }
    function gr(e, r) {
      {
        var l = function() {
          $e || ($e = !0, S("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        l.isReactWarning = !0, Object.defineProperty(e, "key", {
          get: l,
          configurable: !0
        });
      }
    }
    function wr(e, r) {
      {
        var l = function() {
          Ue || (Ue = !0, S("%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        l.isReactWarning = !0, Object.defineProperty(e, "ref", {
          get: l,
          configurable: !0
        });
      }
    }
    var mr = function(e, r, l, h, b, x, _) {
      var E = {
        // This tag allows us to uniquely identify this as a React Element
        $$typeof: n,
        // Built-in properties that belong on the element
        type: e,
        key: r,
        ref: l,
        props: _,
        // Record the component responsible for creating this element.
        _owner: x
      };
      return E._store = {}, Object.defineProperty(E._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: !1
      }), Object.defineProperty(E, "_self", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: h
      }), Object.defineProperty(E, "_source", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: b
      }), Object.freeze && (Object.freeze(E.props), Object.freeze(E)), E;
    };
    function Er(e, r, l, h, b) {
      {
        var x, _ = {}, E = null, k = null;
        l !== void 0 && (Ye(l), E = "" + l), yr(r) && (Ye(r.key), E = "" + r.key), pr(r) && (k = r.ref, hr(r, b));
        for (x in r)
          ue.call(r, x) && !vr.hasOwnProperty(x) && (_[x] = r[x]);
        if (e && e.defaultProps) {
          var P = e.defaultProps;
          for (x in P)
            _[x] === void 0 && (_[x] = P[x]);
        }
        if (E || k) {
          var D = typeof e == "function" ? e.displayName || e.name || "Unknown" : e;
          E && gr(_, D), k && wr(_, D);
        }
        return mr(e, E, k, b, h, te.current, _);
      }
    }
    var Ee = N.ReactCurrentOwner, Me = N.ReactDebugCurrentFrame;
    function Q(e) {
      if (e) {
        var r = e._owner, l = le(e.type, e._source, r ? r.type : null);
        Me.setExtraStackFrame(l);
      } else
        Me.setExtraStackFrame(null);
    }
    var _e;
    _e = !1;
    function be(e) {
      return typeof e == "object" && e !== null && e.$$typeof === n;
    }
    function Ve() {
      {
        if (Ee.current) {
          var e = j(Ee.current.type);
          if (e)
            return `

Check the render method of \`` + e + "`.";
        }
        return "";
      }
    }
    function _r(e) {
      {
        if (e !== void 0) {
          var r = e.fileName.replace(/^.*[\\\/]/, ""), l = e.lineNumber;
          return `

Check your code at ` + r + ":" + l + ".";
        }
        return "";
      }
    }
    var qe = {};
    function br(e) {
      {
        var r = Ve();
        if (!r) {
          var l = typeof e == "string" ? e : e.displayName || e.name;
          l && (r = `

Check the top-level render call using <` + l + ">.");
        }
        return r;
      }
    }
    function Ke(e, r) {
      {
        if (!e._store || e._store.validated || e.key != null)
          return;
        e._store.validated = !0;
        var l = br(r);
        if (qe[l])
          return;
        qe[l] = !0;
        var h = "";
        e && e._owner && e._owner !== Ee.current && (h = " It was passed a child from " + j(e._owner.type) + "."), Q(e), S('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.', l, h), Q(null);
      }
    }
    function Be(e, r) {
      {
        if (typeof e != "object")
          return;
        if (we(e))
          for (var l = 0; l < e.length; l++) {
            var h = e[l];
            be(h) && Ke(h, r);
          }
        else if (be(e))
          e._store && (e._store.validated = !0);
        else if (e) {
          var b = g(e);
          if (typeof b == "function" && b !== e.entries)
            for (var x = b.call(e), _; !(_ = x.next()).done; )
              be(_.value) && Ke(_.value, r);
        }
      }
    }
    function Rr(e) {
      {
        var r = e.type;
        if (r == null || typeof r == "string")
          return;
        var l;
        if (typeof r == "function")
          l = r.propTypes;
        else if (typeof r == "object" && (r.$$typeof === s || // Note: Memo only checks outer props here.
        // Inner props are checked in the reconciler.
        r.$$typeof === c))
          l = r.propTypes;
        else
          return;
        if (l) {
          var h = j(r);
          lr(l, e.props, "prop", h, e);
        } else if (r.PropTypes !== void 0 && !_e) {
          _e = !0;
          var b = j(r);
          S("Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?", b || "Unknown");
        }
        typeof r.getDefaultProps == "function" && !r.getDefaultProps.isReactClassApproved && S("getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.");
      }
    }
    function Tr(e) {
      {
        for (var r = Object.keys(e.props), l = 0; l < r.length; l++) {
          var h = r[l];
          if (h !== "children" && h !== "key") {
            Q(e), S("Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.", h), Q(null);
            break;
          }
        }
        e.ref !== null && (Q(e), S("Invalid attribute `ref` supplied to `React.Fragment`."), Q(null));
      }
    }
    function ze(e, r, l, h, b, x) {
      {
        var _ = O(e);
        if (!_) {
          var E = "";
          (e === void 0 || typeof e == "object" && e !== null && Object.keys(e).length === 0) && (E += " You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");
          var k = _r(b);
          k ? E += k : E += Ve();
          var P;
          e === null ? P = "null" : we(e) ? P = "array" : e !== void 0 && e.$$typeof === n ? (P = "<" + (j(e.type) || "Unknown") + " />", E = " Did you accidentally export a JSX literal instead of a component?") : P = typeof e, S("React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s", P, E);
        }
        var D = Er(e, r, l, b, x);
        if (D == null)
          return D;
        if (_) {
          var M = r.children;
          if (M !== void 0)
            if (h)
              if (we(M)) {
                for (var Z = 0; Z < M.length; Z++)
                  Be(M[Z], e);
                Object.freeze && Object.freeze(M);
              } else
                S("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");
            else
              Be(M, e);
        }
        return e === u ? Tr(D) : Rr(D), D;
      }
    }
    function Sr(e, r, l) {
      return ze(e, r, l, !0);
    }
    function Cr(e, r, l) {
      return ze(e, r, l, !1);
    }
    var xr = Cr, Or = Sr;
    ae.Fragment = u, ae.jsx = xr, ae.jsxs = Or;
  }()), ae;
}
process.env.NODE_ENV === "production" ? Te.exports = Ar() : Te.exports = Lr();
var Wr = Te.exports, Yr = { exports: {} };
(function(t) {
  var n = Object.prototype.hasOwnProperty, o = "~";
  function u() {
  }
  Object.create && (u.prototype = /* @__PURE__ */ Object.create(null), new u().__proto__ || (o = !1));
  function p(s, i, a) {
    this.fn = s, this.context = i, this.once = a || !1;
  }
  function d(s, i, a, c, w) {
    if (typeof a != "function")
      throw new TypeError("The listener must be a function");
    var y = new p(a, c || s, w), m = o ? o + i : i;
    return s._events[m] ? s._events[m].fn ? s._events[m] = [s._events[m], y] : s._events[m].push(y) : (s._events[m] = y, s._eventsCount++), s;
  }
  function v(s, i) {
    --s._eventsCount === 0 ? s._events = new u() : delete s._events[i];
  }
  function f() {
    this._events = new u(), this._eventsCount = 0;
  }
  f.prototype.eventNames = function() {
    var i = [], a, c;
    if (this._eventsCount === 0)
      return i;
    for (c in a = this._events)
      n.call(a, c) && i.push(o ? c.slice(1) : c);
    return Object.getOwnPropertySymbols ? i.concat(Object.getOwnPropertySymbols(a)) : i;
  }, f.prototype.listeners = function(i) {
    var a = o ? o + i : i, c = this._events[a];
    if (!c)
      return [];
    if (c.fn)
      return [c.fn];
    for (var w = 0, y = c.length, m = new Array(y); w < y; w++)
      m[w] = c[w].fn;
    return m;
  }, f.prototype.listenerCount = function(i) {
    var a = o ? o + i : i, c = this._events[a];
    return c ? c.fn ? 1 : c.length : 0;
  }, f.prototype.emit = function(i, a, c, w, y, m) {
    var T = o ? o + i : i;
    if (!this._events[T])
      return !1;
    var g = this._events[T], N = arguments.length, S, C;
    if (g.fn) {
      switch (g.once && this.removeListener(i, g.fn, void 0, !0), N) {
        case 1:
          return g.fn.call(g.context), !0;
        case 2:
          return g.fn.call(g.context, a), !0;
        case 3:
          return g.fn.call(g.context, a, c), !0;
        case 4:
          return g.fn.call(g.context, a, c, w), !0;
        case 5:
          return g.fn.call(g.context, a, c, w, y), !0;
        case 6:
          return g.fn.call(g.context, a, c, w, y, m), !0;
      }
      for (C = 1, S = new Array(N - 1); C < N; C++)
        S[C - 1] = arguments[C];
      g.fn.apply(g.context, S);
    } else {
      var q = g.length, U;
      for (C = 0; C < q; C++)
        switch (g[C].once && this.removeListener(i, g[C].fn, void 0, !0), N) {
          case 1:
            g[C].fn.call(g[C].context);
            break;
          case 2:
            g[C].fn.call(g[C].context, a);
            break;
          case 3:
            g[C].fn.call(g[C].context, a, c);
            break;
          case 4:
            g[C].fn.call(g[C].context, a, c, w);
            break;
          default:
            if (!S)
              for (U = 1, S = new Array(N - 1); U < N; U++)
                S[U - 1] = arguments[U];
            g[C].fn.apply(g[C].context, S);
        }
    }
    return !0;
  }, f.prototype.on = function(i, a, c) {
    return d(this, i, a, c, !1);
  }, f.prototype.once = function(i, a, c) {
    return d(this, i, a, c, !0);
  }, f.prototype.removeListener = function(i, a, c, w) {
    var y = o ? o + i : i;
    if (!this._events[y])
      return this;
    if (!a)
      return v(this, y), this;
    var m = this._events[y];
    if (m.fn)
      m.fn === a && (!w || m.once) && (!c || m.context === c) && v(this, y);
    else {
      for (var T = 0, g = [], N = m.length; T < N; T++)
        (m[T].fn !== a || w && !m[T].once || c && m[T].context !== c) && g.push(m[T]);
      g.length ? this._events[y] = g.length === 1 ? g[0] : g : v(this, y);
    }
    return this;
  }, f.prototype.removeAllListeners = function(i) {
    var a;
    return i ? (a = o ? o + i : i, this._events[a] && v(this, a)) : (this._events = new u(), this._eventsCount = 0), this;
  }, f.prototype.off = f.prototype.removeListener, f.prototype.addListener = f.prototype.on, f.prefixed = o, f.EventEmitter = f, t.exports = f;
})(Yr);
var G;
(function(t) {
  t.Installed = "Installed", t.NotDetected = "NotDetected", t.Loadable = "Loadable", t.Unsupported = "Unsupported";
})(G || (G = {}));
class z extends Error {
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  constructor(o, u) {
    super(o);
    Y(this, "error");
    this.error = u;
  }
}
class $r extends z {
  constructor() {
    super(...arguments);
    Y(this, "name", "WalletNotReadyError");
  }
}
class Ce extends z {
  constructor() {
    super(...arguments);
    Y(this, "name", "WalletNotSelectedError");
  }
}
class X extends z {
  constructor() {
    super(...arguments);
    Y(this, "name", "WalletNotConnectedError");
  }
}
var ve;
(function(t) {
  t.Testnet = "testnet3";
})(ve || (ve = {}));
var pe;
(function(t) {
  t.NoDecrypt = "NO_DECRYPT", t.UponRequest = "DECRYPT_UPON_REQUEST", t.AutoDecrypt = "AUTO_DECRYPT", t.OnChainHistory = "ON_CHAIN_HISTORY";
})(pe || (pe = {}));
class Ur {
  constructor(n, o, u) {
    Y(this, "program");
    Y(this, "functionName");
    Y(this, "inputs");
    this.program = n, this.functionName = o, this.inputs = u;
  }
}
class tr {
  constructor(n, o, u, p, d = !0) {
    Y(this, "address");
    Y(this, "chainId");
    Y(this, "transitions");
    Y(this, "fee");
    Y(this, "feePrivate");
    this.address = n, this.chainId = o, this.transitions = u, this.fee = p, this.feePrivate = d;
  }
  static createTransaction(n, o, u, p, d, v, f = !0) {
    const s = new Ur(u, p, d);
    return new tr(n, o, [s], v, f);
  }
}
function Mr(t, n) {
  const o = R(() => {
    try {
      const d = localStorage.getItem(t);
      if (d)
        return JSON.parse(d);
    } catch (d) {
      typeof window != "undefined" && console.error(d);
    }
    return n;
  }), u = o[0], p = Se(!0);
  return K(() => {
    if (p.current) {
      p.current = !1;
      return;
    }
    try {
      u === null ? localStorage.removeItem(t) : localStorage.setItem(t, JSON.stringify(u));
    } catch (d) {
      typeof window != "undefined" && console.error(d);
    }
  }, [u, t]), A(() => o, [o]);
}
const Vr = {
  selectedWalletName: null,
  setSelectedWalletName: (t) => console.error(de("get", "setSelectedWalletName")),
  connecting: !1,
  setConnecting: (t) => console.error(de("get", "setConnecting")),
  disconnecting: !1,
  setDisconnecting: (t) => console.error(de("get", "setDisconnecting")),
  decryptPermission: pe.NoDecrypt,
  network: ve.Testnet,
  programs: [],
  readyState: G.Unsupported,
  connected: !1,
  handleError: (t) => console.error(de("get", "handleError")),
  publicKey: null,
  wallet: null,
  wallets: [],
  adapter: null
};
function de(t, n) {
  return `You have tried to  ${t} "${n}" on a WalletContext without providing one. Make sure to render a WalletProvider as an ancestor of the component that uses WalletContext`;
}
const nr = Ir(
  Vr
);
function V() {
  return Fr(nr);
}
const er = {
  wallet: null,
  adapter: null,
  publicKey: null,
  connected: !1
}, zr = ({
  children: t,
  wallets: n,
  autoConnect: o = !1,
  decryptPermission: u = pe.NoDecrypt,
  network: p = ve.Testnet,
  onError: d,
  localStorageKey: v = "walletName",
  programs: f = []
}) => {
  const [s, i] = Mr(
    v,
    null
  ), [{ wallet: a, adapter: c, publicKey: w, connected: y }, m] = R(er), T = (c == null ? void 0 : c.readyState) || G.Unsupported, [g, N] = R(!1), [S, C] = R(!1), q = Se(!1), [U, se] = R(
    () => n.map((O) => ({
      adapter: O,
      readyState: O.readyState
    }))
  );
  K(() => {
    se(
      (L) => n.map((W, j) => {
        const F = L[j];
        return F && F.adapter === W && F.readyState === W.readyState ? F : {
          adapter: W,
          readyState: W.readyState
        };
      })
    );
    function O(L) {
      se((W) => {
        const j = W.findIndex(({ adapter: H }) => H === this);
        if (j === -1)
          return W;
        const { adapter: F } = W[j];
        return [
          ...W.slice(0, j),
          { adapter: F, readyState: L },
          ...W.slice(j + 1)
        ];
      });
    }
    return n.forEach(
      (L) => L.on("readyStateChange", O, L)
    ), () => n.forEach(
      (L) => L.off("readyStateChange", O, L)
    );
  }, [n]), K(() => {
    const O = s && U.find(({ adapter: L }) => L.name === s);
    m(O ? {
      wallet: O,
      adapter: O.adapter,
      connected: O.adapter.connected,
      publicKey: O.adapter.publicKey
    } : er);
  }, [s, U]), K(() => {
    function O() {
      q.current = !0;
    }
    return window.addEventListener("beforeunload", O), () => window.removeEventListener("beforeunload", O);
  }, [q]);
  const ee = I(() => {
    c && m((O) => Xe(He({}, O), {
      connected: c.connected,
      publicKey: c.publicKey
    }));
  }, [c]), re = I(() => {
    q.current || i(null);
  }, [q, i]), J = I(
    (O) => (q.current || (d || console.error)(O), O),
    [q, d]
  );
  return K(() => {
    if (c)
      return c.on("connect", ee), c.on("disconnect", re), c.on("error", J), () => {
        c.off("connect", ee), c.off("disconnect", re), c.off("error", J);
      };
  }, [c, ee, re, J]), K(() => () => {
    c == null || c.disconnect();
  }, [c]), K(() => {
    S || y || !o || !c || !(T === G.Installed || T === G.Loadable) || function() {
      return $(this, null, function* () {
        C(!0);
        try {
          yield c.connect(u, p, f);
        } catch (O) {
          i(null);
        } finally {
          C(!1);
        }
      });
    }();
  }, [S, y, o, c, T, i]), A(
    () => /* @__PURE__ */ Wr.jsx(
      nr.Provider,
      {
        value: {
          selectedWalletName: s,
          setSelectedWalletName: i,
          connecting: S,
          setConnecting: C,
          disconnecting: g,
          setDisconnecting: N,
          decryptPermission: u,
          network: p,
          programs: f,
          readyState: T,
          connected: y,
          handleError: J,
          publicKey: w,
          wallet: a,
          wallets: U,
          adapter: c
        },
        children: t
      }
    ),
    [
      s,
      i,
      S,
      C,
      g,
      N,
      u,
      p,
      f,
      T,
      y,
      J,
      w,
      a,
      U,
      c,
      t
    ]
  );
}, Jr = () => {
  const { publicKey: t, connected: n } = V();
  return A(() => ({ publicKey: t, connected: n }), [t, n]);
}, Gr = ({ program: t = "credits.aleo", enabled: n = !0 } = {
  program: "credits.aleo",
  enabled: !0
}) => {
  const [o, u] = R(null), [p, d] = R(null), [v, f] = R(!1), { connected: s, adapter: i } = V(), a = I(
    (c) => $(void 0, null, function* () {
      f(!0);
      try {
        if (i && "requestRecords" in i) {
          if (!s)
            throw new X();
          const w = yield i.requestRecords(t);
          if (c) {
            const y = w.filter(c);
            return u(y), y;
          }
          return u(w), w;
        } else
          throw new z("Not implemented");
      } catch (w) {
        d(w);
      } finally {
        f(!1);
      }
    }),
    [t, i, s]
  );
  return K(() => {
    n && a();
  }, [n, i]), A(
    () => ({ records: o, error: p, loading: v, requestRecords: a }),
    [o, p, v, a]
  );
};
var oe = /* @__PURE__ */ ((t) => (t.DownloadingProverFiles = "Downloading Prover Files", t.GeneratingTransaction = "Generating Transaction", t.Broadcasting = "Broadcasting", t.Failed = "Failed", t.Completed = "Completed", t.Finalized = "Finalized", t))(oe || {});
const qr = () => {
  const { connected: t, adapter: n } = V(), [o, u] = R(null), [p, d] = R(!1), [v, f] = R(null), s = I(
    (i) => $(void 0, null, function* () {
      try {
        if (d(!0), !t)
          throw new X();
        if (n && "transactionStatus" in n) {
          const a = yield n.transactionStatus(i);
          return f(a), a;
        } else
          throw new z("Not implemented in your wallet provider");
      } catch (a) {
        return u(a), null;
      } finally {
        d(!1);
      }
    }),
    [t, n]
  );
  return A(
    () => ({ requestTransactionStatus: s, error: o, transactionStatus: v, loading: p }),
    [s, o, v, p]
  );
}, Hr = ({ poolingInterval: t = 1e3 } = { poolingInterval: 1e3 }) => {
  const { requestTransactionStatus: n } = qr(), [o, u] = R(null), [p, d] = R(null), v = Se(null), f = I(() => {
    v.current && clearInterval(v.current);
  }, [v]), s = I(
    (i, a) => $(void 0, null, function* () {
      return d(null), yield new Promise((c, w) => {
        try {
          v.current && clearInterval(v.current), n && (v.current = setInterval(() => $(void 0, null, function* () {
            const y = yield n(i);
            u(y), a && a(y, f), (y === oe.Failed || y === oe.Finalized) && v.current && (clearInterval(v.current), y === oe.Failed && w(y), y === oe.Finalized && c(y));
          }), t));
        } catch (y) {
          d(y), w(y);
        }
      });
    }),
    [t, n]
  );
  return A(() => ({ status: o, error: p, wait: s }), [o, p, s]);
}, Xr = () => {
  const {
    connecting: t,
    setConnecting: n,
    disconnecting: o,
    connected: u,
    handleError: p,
    setSelectedWalletName: d,
    decryptPermission: v,
    network: f,
    programs: s,
    wallets: i,
    publicKey: a
  } = V(), [c, w] = R(null), y = I(
    (m) => $(void 0, null, function* () {
      if (t || o || u)
        return;
      const T = i.find((g) => g.adapter.name === m);
      if (!(T != null && T.adapter))
        throw p(new Ce());
      if (!(T.readyState === G.Installed || T.readyState === G.Loadable))
        throw d(null), typeof window != "undefined" && window.open(T.adapter.url, "_blank"), p(new $r());
      n(!0);
      try {
        yield T.adapter.connect(v, f, s), d(m);
      } catch (g) {
        d(null), w(g);
      } finally {
        n(!1);
      }
    }),
    [t, o, u, i, p, d]
  );
  return A(
    () => ({ connected: u, connecting: t, connect: y, error: c, address: a }),
    [u, t, y, c, a]
  );
}, Qr = () => {
  const { disconnecting: t, setDisconnecting: n, adapter: o, setSelectedWalletName: u } = V(), [p, d] = R(null), v = I(() => $(void 0, null, function* () {
    if (!t) {
      if (!o)
        return u(null);
      n(!0);
      try {
        yield o.disconnect();
      } catch (f) {
        u(null), d(f);
      } finally {
        n(!1);
      }
    }
  }), [t, o, u]);
  return A(() => ({ disconnect: v, disconnecting: t, error: p }), [v, t, p]);
}, Zr = () => {
  const { setSelectedWalletName: t } = V(), n = I((o) => {
    t(o);
  }, []);
  return A(() => ({ select: n }), [n]);
}, et = ({
  cipherText: t,
  tpk: n,
  programId: o,
  functionName: u,
  index: p,
  enabled: d = !0
}) => {
  const [v, f] = R(null), [s, i] = R(d), [a, c] = R(null), { adapter: w, connected: y } = V(), m = I(() => $(void 0, null, function* () {
    i(!0), c(null);
    try {
      if (!y)
        throw new X();
      if (!w)
        throw new Ce();
      if (w && "decrypt" in w) {
        const T = yield w.decrypt(
          t,
          n,
          o,
          u,
          p
        );
        f(T);
      }
    } catch (T) {
      c(T);
    } finally {
      i(!1);
    }
  }), [y, w, t, n, o, u, p]);
  return K(() => {
    d && m();
  }, [m, d]), A(
    () => ({ decryptedText: v, loading: s, error: a, decrypt: m }),
    [v, s, a, m]
  );
}, rt = ({ enabled: t } = { enabled: !0 }) => {
  const { adapter: n, connected: o } = V(), [u, p] = R(null), [d, v] = R(t), [f, s] = R(null), i = I(() => $(void 0, null, function* () {
    v(!0), s(null);
    try {
      if (!o)
        throw new X();
      if (!n)
        throw new Ce();
      if (n && "requestViewKey" in n) {
        const a = yield n.requestViewKey();
        p(a);
      } else
        throw new z("Not implemented in your wallet provider");
    } catch (a) {
      s(a);
    } finally {
      v(!1);
    }
  }), [o, n]);
  return K(() => {
    t && i();
  }, [i]), A(
    () => ({ viewKey: u, loading: d, error: f, requestViewKey: i }),
    [u, d, f, i]
  );
}, tt = () => {
  const { connected: t, adapter: n } = V(), [o, u] = R(null), [p, d] = R(!1), [v, f] = R(null), s = I(
    (i) => $(void 0, null, function* () {
      try {
        if (d(!0), !t)
          throw new X();
        if (n && "requestBulkTransactions" in n) {
          const a = yield n.requestBulkTransactions(i);
          return f(a), a;
        } else
          throw new z("Not implemented in your wallet provider");
      } catch (a) {
        return u(a), null;
      } finally {
        d(!1);
      }
    }),
    [n]
  );
  return A(
    () => ({ executeBulkTransactions: s, error: o, transactionIds: v, loading: p }),
    [s, o, v, p]
  );
}, nt = () => {
  const { connected: t, adapter: n } = V(), [o, u] = R(null), [p, d] = R(!1), [v, f] = R(null), s = I((i) => $(void 0, null, function* () {
    try {
      if (d(!0), !t)
        throw new X();
      if (n && "requestTransaction" in n) {
        const a = yield n.requestTransaction(i);
        return f(a), a;
      } else
        throw new z("Not implemented in your wallet provider");
    } catch (a) {
      return u(a), null;
    } finally {
      d(!1);
    }
  }), []);
  return A(
    () => ({ executeTransaction: s, error: o, transactionId: v, loading: p }),
    [s, o, v, p]
  );
}, at = () => {
  const { connected: t, adapter: n } = V(), [o, u] = R(null), [p, d] = R(!1), [v, f] = R(null), s = I(
    (i) => $(void 0, null, function* () {
      try {
        if (d(!0), !t)
          throw new X();
        if (n && "signMessage" in n) {
          const a = yield n.signMessage(i);
          return f(a), a;
        } else
          throw new z("Not implemented in your wallet provider");
      } catch (a) {
        return u(a), null;
      } finally {
        d(!1);
      }
    }),
    [t, n]
  );
  return A(
    () => ({ signMessage: s, error: o, data: v, loading: p }),
    [s, o, v, p]
  );
};
export {
  pe as DecryptPermission,
  tr as Transaction,
  oe as TransactionStatus,
  ve as WalletAdapterNetwork,
  zr as WalletProvider,
  Jr as useAccount,
  tt as useBulkTransactions,
  Xr as useConnect,
  et as useDecrypt,
  Qr as useDisconnect,
  Gr as useRecords,
  Zr as useSelect,
  at as useSignMessage,
  nt as useTransaction,
  qr as useTransactionStatus,
  rt as useViewKey,
  Hr as useWait
};
